### PR TITLE
Integrate CoreFoundation with the asyncio event loop.

### PR DESCRIPTION
**Pull request by freakboy3742**

### CODE REVIEW

**Comment by dgelessus** [comment]
Right, Travis is still broken on the main branch. There's a fix for the Travis issue in one commit that's part of #76, but I haven't had time to finish that yet...

Would it be possible to implement this using Foundation APIs rather than Core Foundation? IMHO it would be better if we eventually migrated away from CF completely, see #74.

I'm not very familiar with async stuff, so I can't really comment on much else, sorry.

**Comment by freakboy3742** [comment]
Can be done with Foundation APIs... I'm not sure. I'm mostly stumbling around trying to work out how Python's asyncio integration works, and what minimal examples exist in this space all seem to use the CoreFoundation libraries. 

At a guess, this is because integrating with asyncio requires fairly low level API access, and some of the object lifecycle handling around the Foundation classes can get in the way. For example, rewriting what is currently on this branch to use `NSTimer` instead of `CFRunLoopTimerCreate` does simplify the initial calls; but you lose the fine grained control over timer lifecycle and execution time.

However, it may also be because integrating with C libraries is easier than integrating with Objective C (unless, of course, you've got Rubicon backing you up!)

At the very least, it should be possible to restrict CoreFoundation usage to just the async module within Rubicon. For at least this first pass, I'd like to stick to CoreFoundation; once I've got that working, we can look at dropping CoreFoundation in revised versions.

**Comment by dgelessus** [comment]
Sounds reasonable. Now that I think about it, I remember reading that some types aren't toll-free-bridged even though they have similar names. `CFRunLoop`/`NSRunLoop` might be one of those cases.

Just to be clear - are you still working on the PR, or is it finished? Because the commit says "First pieces" it sounds like there are other parts that you want to add later. (There are also no tests and docs yet.)

**Comment by freakboy3742** [comment]
Sorry - I forgot to apply a label. Yes, this is a work in progress. I've only got timers working so far, not the rest of communications, and as you've noted, there aren't any docs or tests yet. I'm working on those at the moment.

**Comment by dgelessus** [comment]
No worries :) Just wanted to check so that I don't merge it too early.

File: rubicon/objc/async.py [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+        self._writer = (callback, args)
+
+    def disable_write(self):
+        libcf.CFSocketDisableCallBacks(self._cf_socket, kCFSocketWriteCallBack)
+        self._writer = None
+        self.cancel()
+
+    def cancel(self):
+        """
+        """
```
Comment (dgelessus) [inline rubicon/objc/async.py]: This should probably be removed (or filled).

**Comment by dgelessus** [comment]
Late question: what is the syntax that you're using in the docstrings? I've seen it used in other places already, but I don't know what tool it's from.

Would it make sense to use Sphinx syntax in docstrings, so we could auto-generate reference docs from docstrings later? Sphinx seems to support this: https://pythonhosted.org/an_example_pypi_project/sphinx.html#auto-directives

File: rubicon/objc/async.py (hunk 1 above)
Comment (freakboy3742) [reply rubicon/objc/async.py]: Unfortunately, this comment has been cancelled. :-)

File: rubicon/objc/async.py [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+    _fields_ = [
+        ('copyDescription', CFUNCTYPE(CFStringRef, c_void_p)),  # CFStringRef (*copyDescription)(const void *info)
+        ('info', c_void_p),
+        ('release', CFUNCTYPE(None, c_void_p)),  # void (*release)(const void *info)
+        ('retain', CFUNCTYPE(None, c_void_p)),  # const void *(*retain)(const void *info)
+        ('version', CFIndex),
+    ]
+
+
+kCFAllocatorDefault = objc_const(libcf, 'kCFAllocatorDefault')
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply rubicon/objc/async.py]: I didn't see this one in the Apple docs. Agreed that using None directly makes more sense.

File: rubicon/objc/async.py [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+from .runtime import objc_const, objc_id
+from .core_foundation import CFStringRef, libcf
+
+###########################################################################
+# CoreFoundation types and constants needed for async handlers
+###########################################################################
+CFAllocatorRef = objc_id
+CFData = objc_id
+CFIndex = c_long
+CFOptionFlags = c_ulong
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply rubicon/objc/async.py]: Good point. I'll move and/or reuse the more generic ones.

**Comment by freakboy3742** [comment]
@dgelessus I'm not entirely sure what you're referring to with the doctoring format. I've been pretty much doing the "first line is a summary, subsequent lines are details" format, without any specific directives. In many cases, the comment has been lifted directly from the CPython implementation (since that definition is still accurate).

I know Toga has been going a bit more detailed with argument definitions, specifically so it can be rendered as autodoc - but I haven't been as focussed on this for Rubicon. Do you think it's worth adding?

**Comment by freakboy3742** [comment]
Ok - this is pretty much ready now. It's tested (as much as possible), it has docs, and it has a placeholder for iOS support once I work out how that will work. Pending your final review, @dgelessus, I think we're ready to go.

**Comment by dgelessus** [comment]
Sorry, I should have been more specific - I was referring to [the docstring of `CFSocketHandle.__init__`](https://github.com/pybee/rubicon-objc/pull/78/files#diff-4567f90e10dd143f8435e519e7257d7bR203) and the `C{...}` and `L{...}` markup used there. (I thought I had seen the markup in more places, so I didn't point to a specific docstring, but that one seems to be the only one.)

I think autodoc would be useful for filling out the "reference" section of the docs - right now we only have the tutorials and how-tos. While the how-tos are great when using a feature for the first time, I think we should also have a reference generated from the docstrings where all constants, functions and classes are documented by module. That way when you (for example) see `ObjCClass.declare_property` used in some code, you can go into the reference docs for that method and look up what it does, without having to find the how-to page in which it is explained. And since the reference docs are generated from docstrings, you can also use `help(ObjCClass.declare_property)` in the REPL.

**Comment by freakboy3742** [comment]
Ah! Yeah - that docstring is Twisted's docs format - and I agree that Sphinx's is better. I'll clean that one up.

**Pull request merged**